Remove ComponentLifecycle interface in favour of individual interfaces

I'd like to propose removing the [ComponentLifecycle](https://github.com/developit/preact/blob/master/src/preact.d.ts#L25-L34) interface and replace it with individual interfaces, e.g.
```ts
interface WillMount {
    componentWillMount(): void;
}

interface WillReceiveProps<PropsType, ContextType> {
    componentWillReceiveProps(nextProps: PropsType, nextContext: ContextType): void;
}
```

This way instead of Component implementing the ComponentLifecycle interface where everything is marked optional, the consumer would be able to explicitly implement WillMount, for example.

This allows for greater type safety as TypeScript will be able to know if the interface is implemented incorrectly, something it can't do with the current solution.
![](https://i.uwf.io/nwaIQg.png)

This has the added benefit of allowing vscode (and presumably other editors) to auto implement the interfaces. (And I wouldn't have to flick to the docs because I forgot the signature literally every time I need to implement a lifecycle event 😳)

I realise this would deviate from the React typings, but I am hoping that is not a major concern.

Unless I'm missing something, splitting up the interface shouldn't be necessary to get type safety. Since JavaScript does not support overloaded methods (ie. there can only be one implementation with a given name), then TypeScript knows that if a particular optional method is implemented then it must conform to the signature of any method with the same name in implemented interfaces.

```ts
import { Component } from 'preact';

class Thing extends Component<any,any> implements preact.ComponentLifecycle<any,any> {
       // error TS2420: Class 'Thing' incorrectly implements interface 'ComponentLifecycle<any, any>'
	componentWillMount(a: string): number {
		return 42;
	}

	render() {
		return null;
	}
}
```

One slightly surprising thing I noticed here is that `implements` is not transitive, so although `Component` is declared to implement `ComponentLifecycle`, subclasses of `Component` are not automatically required to implement it.

--- a/src/preact.d.ts
+++ b/src/preact.d.ts
@@ -68,13 +68,13 @@ declare namespace preact {
 	type AnyComponent<PropsType, StateType> = FunctionalComponent<PropsType> | typeof Component;
 
 	interface Component<PropsType, StateType> {
-		componentWillMount(): void;
-		componentDidMount(): void;
-		componentWillUnmount(): void;
-		componentWillReceiveProps(nextProps: Readonly<PropsType>, nextContext: any): void;
-		shouldComponentUpdate(nextProps: Readonly<PropsType>, nextState: Readonly<StateType>, nextContext: any): boolean;
-		componentWillUpdate(nextProps: Readonly<PropsType>, nextState: Readonly<StateType>, nextContext: any): void;
-		componentDidUpdate(previousProps: Readonly<PropsType>, previousState: Readonly<StateType>, previousContext: any): void;
+		componentWillMount?(): void;
+		componentDidMount?(): void;
+		componentWillUnmount?(): void;
+		componentWillReceiveProps?(nextProps: Readonly<PropsType>, nextContext: any): void;
+		shouldComponentUpdate?(nextProps: Readonly<PropsType>, nextState: Readonly<StateType>, nextContext: any): boolean;
+		componentWillUpdate?(nextProps: Readonly<PropsType>, nextState: Readonly<StateType>, nextContext: any): void;
+		componentDidUpdate?(previousProps: Readonly<PropsType>, previousState: Readonly<StateType>, previousContext: any): void;
 	}
 
 	abstract class Component<PropsType, StateType> {

